keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,750 @@
+       enum: ["EmptyDir","HostPath"]
+     emptyDir:
+       type: object
+     hostPath:
+       type: object
+       properties:
+         path:
+           type: string
+           pattern: "/.*"
+         required: ["path"]
+     sharedField:
+       type: string
+   required:
+      - sharedField
+   ```
+   and published with the following additional schema fields:
+   ```yaml
+   oneOf:
+   - required: ["emptyDir"]
+   - required: ["hostPath"]
+   ```
+   After dropping the value validation, we get:
+   ```yaml
+   type: object
+   x-kubernetes-union-field-values:
+     emptyDir: EmptyDir
+     hostPath: HostPath
+   x-kubernetes-union-discriminator: type
+   properties:
+     type:
+       type: string
+     emptyDir:
+       type: object
+     hostPath:
+       type: object
+       properties:
+         path:
+           type: string
+     sharedField:
+       type: string
+   ```
+3. non-polymorphic type, polymorphic format for IPv4 and IPv6:
+   ```go
+   // +kubebuilder:anyOf:pattern="^(?:(?:25[0...$",pattern="^(([0-9a-fA-F]...$"
+   type IPv4Orv6 string
+   ```
+   is specified by:
+   ```yaml
+   type: string
+   anyOf:
+   - pattern: "^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}...$"
+   - pattern: "(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|...$"
+   ```
+   and published the same way.
+  
+   After dropping the value validation, we get:
+   ```yaml
+   type: string
+   ```
+
+4. RawExtensions, i.e. unconstrained, embedded toplevel types 
+   ```go
+   type Foo struct {
+     // +nullable
+     // +kubebuilder:properties=apiVersion:pattern=v1
+     X runtime.RawExtensions
+   }
+   ```
+   specified by:
+   ```yaml
+   type: object
+   nullable: true
+   x-kubernetes-embedded-resource: true
+   x-kubernetes-preserve-unknown-fields: true
+   properties:
+     apiVersion:
+       pattern: v1
+   ```
+   and published with the following additional schema fields:
+   ```yaml
+   properties:
+     apiVersion:
+       allOf:
+       - type: string
+         pattern: <ome-apiVersion-regex>
+       - pattern: v1
+     kind:
+       type: string
+     metadata:
+       type: object
+       <some-further-ObjectMeta-schema>
+   ```
+   After dropping the value validation, we get:
+   ```yaml
+   type: object
+   nullable: true
+   x-kubernetes-embedded-resource: true
+   x-kubernetes-preserve-unknown-fields: true
+   properties:
+     apiVersion:
+       type: string
+     kind:
+       type: string
+     metadata:
+       type: object
+       <some-further-non-value-validation-ObjectMeta-schema>
+   ```
+
+5. embedded JSON, further specified by the user using tags:
+   ```go
+   type Foo struct {
+     // +nullable
+     // +kubebuilder:type=object
+     Json JSON
+   }
+   ```
+   specified by:
+   ```yaml
+   type: object
+   properties:
+     json:
+       type: object
+       x-kubernetes-preserve-unknown-fields: true
+       nullable: true
+   ```
+   and published the same way. Dropping value validation does not change anything.
```

Comment:
can you clarify whether the "published" examples here refer to openapi v2 or v3? things like `nullable` don't exist in v2, right?

Reply:
clarified by giving both OpenAPI v3 and v2 output